src/lib/Embed.hs

```diff
@@ -399,9 +396,35 @@ emitMaybeCase scrut nothingCase justCase = do
   let resultTy = getType nothingBody
   emit $ Case scrut [nothingAlt, justAlt] resultTy
 
-emitRunWriter :: MonadEmbed m => Name -> Type -> (Atom -> m Atom) -> m Atom
-emitRunWriter v ty body = do
-  emit . Hof . RunWriter =<< mkBinaryEffFun Writer v ty body
+monoidLift :: Type -> Type -> Nest Binder
+monoidLift baseTy accTy = case baseTy == accTy of
+  True  -> Empty
+  False -> case accTy of
+    TabTy n b -> Nest n $ monoidLift baseTy b
+    _         -> error $ "Base monoid type mismatch: can't lift " ++
+                         pprint baseTy ++ " to " ++ pprint accTy
+
+mextendForRef :: MonadEmbed m => Atom -> BaseMonoid -> Atom -> m Atom
```

Comment:
I'm realizing I may have misunderstood what this is for. My new understanding:

For simple `runAccum` code without optimization, `(+=)` uses the normal typeclass resolution to figure out how to update whatever type of reference we have, and that update function gets passed to the `%mextend` primitive. In that case, it's entirely up to user code what that means (which is awesome).

However, when converting to an efficient tiled version, we no longer have access to the dictionary synthesis machinery. So we need to build our own updater in the compiler, by lifting from a base instance. That is what this function does. For correctness, this needs to be equivalent to the updater that was used in userspace. But because we wrote both (one in the compiler, one in the prelude) we can guarantee that this is true. In this restricted setting, we only care about tables, because we only tile across tables.

Remaining question: theoretically, if we could run synthesis in `MonadEmbed` as @dougalm suggested, would this no longer be necessary? Because we could just ask for the appropriate `(+=)` from userspace instead of building it separately in the compiler?

Reply:
Exactly, you've figured it out perfectly! And yes, it's possible that we could reuse synthesis for the same purpose. For now I've concluded that this is a little simpler, because we only have a single case, but if we ever wanted to make it more general then that would probably be a good idea. In any case, I'm fine with both options, _maybe_ with a slight preference towards exposing dictionary synthesis here. But that sounds like something to do in a follow up PR (and I don't think I'll have time, but would gladly accept patches if they're simple 😄)

Reply:
Sounds good to do this for now. In principle I like the idea of eventually doing it via dictionary synthesis, since that seems easier to avoid mysterious compiler-doesn't-match-prelude bugs.